packages/edit-site/src/store/actions.js

```diff
@@ -105,21 +105,62 @@ export function setTemplatePart( templatePartId ) {
 }
 
 /**
- * Resolves the template for a page and sets them.
+ * Updates the homeTemplateId state with the templateId of the page resolved
+ * from the given path.
  *
- * @param {Object} page         The page object.
- * @param {string} page.type    The page type.
- * @param {string} page.slug    The page slug.
- * @param {string} page.path    The page path.
- * @param {Object} page.context The page context.
+ * @param {number} homeTemplateId The template ID for the homepage.
+ */
+export function setHomeTemplateId( homeTemplateId ) {
+	return {
+		type: 'SET_HOME_TEMPLATE',
+		homeTemplateId,
+	};
+}
+
+/**
+ * Resolves the template for a page if no templateId is passed and sets them.
+ *
+ * @param {Object}  page         The page object.
+ * @param {string}  page.type    The page type.
+ * @param {string}  page.slug    The page slug.
+ * @param {string}  page.path    The page path.
+ * @param {Object}  page.context The page context.
+ * @param {number?} templateId   The template ID for the page if we already know it.
  *
  * @return {Object} Action object.
  */
-export function* setPage( page ) {
-	const templateId = yield findTemplate( page.path );
+export function* setPage( page, templateId ) {
```

Comment:
From a separation of concerns POV, I'm still not too happy we're passing `templateId` to `setPage`.
Since it's obtained via `findTemplate` both here, and in `showHomepage` -- is there any chance this is being cached at some level anyway?

Reply:
We could alternatively absorb some of the extra logic from `showHomepage` into this function, and run the `show_on_front`/`page_on_front` stuff only `if ( page === '/' )` :thinking:

Reply:
The request does happen twice without this code, but we should definitely try to cache it properly! I'm not sure where to start to make that happen, though.

<img width="1389" alt="Screen Shot 2020-07-13 at 11 39 45 AM" src="https://user-images.githubusercontent.com/6265975/87340923-9ee81000-c4fd-11ea-9316-0df9fc671155.png">

> We could alternatively absorb some of the extra logic from showHomepage into this function

Yeah, that could work too. That means that we aren't always relying solely on the passed `page` object, though.

Alternatively, we could add a `setRoute` function which:
- computes required page context (like query/page ID)
- computes required template ID
- sets the page & template

and then we do `setRoute( '/' );`

But that doesn't solve the problem of passing the templateId to setPage.

We could always let it fetch twice, but it'd be nice to avoid the performance hit

Reply:
Ok, I've added a naive cache layer to `findTemplate`. I think this solves the problem well enough for now (though we might need to think about the consequences a bit more).

Ideally, (per @epiqueras) we would add `path` to the getEntityRecords function. Then it could be passed to the actual WordPress query. The problem here is that it would be fairly complicated to create a `get_template_from_path` function server-side because template resolution is tied so closely with page render.

Reply:
> Ok, I've added a naive cache layer to `findTemplate`. I think this solves the problem well enough for now (though we might need to think about the consequences a bit more).

Thanks! I'm unfortunately still a bit concerned about it living at the module level :sweat_smile: 
We've repeatedly seen issues in other projects that import `@wordpress/` dependencies with some of their singleton-style data management (specifically when different versions of one `@wordpress/` package -- most notably `@wordpress/data` I think -- ended up in our npm dependency tree without getting de-duplicated).

So I'd rather steer clear of module-level caching and the like. I noticed that we're [creating a registry control for the `findTemplate` util](https://github.com/WordPress/gutenberg/blob/1ba985d852dbb2ca8d91befe2b30a552fe251794/packages/edit-site/src/store/controls.js#L26-L31) in [`controls.js`](https://github.com/WordPress/gutenberg/blob/1ba985d852dbb2ca8d91befe2b30a552fe251794/packages/edit-site/src/store/controls.js). I'm not too familiar with controls' (and registries') internals, but this was one of the reasons I was hoping we'd be some caching for free (I do seem to remember that caching for network requests is added in a similar way in `wp/core` or `core-data`.) More importantly, since we're essentially wrapping the util in a control (that is tied to a registry), I was thinking that we _should_ at least be caching at registry level -- I think there can multiple registries, so I think that each should have its individual cache.

My apologies, I truly don't mean to stall this PR. Maybe we should remove the caching for now and add it separately in a subsequent PR? I feel that I personally don't know enough about controls and registries, and that there might be a 'right' way to do it that I have yet to learn about. I'm a bit uncomfortable with the module level caching and would lean towards removing it.

> Ideally, (per @epiqueras) we would add `path` to the getEntityRecords function. Then it could be passed to the actual WordPress query. The problem here is that it would be fairly complicated to create a `get_template_from_path` function server-side because template resolution is tied so closely with page render.

Yep! That'd be the holy grail. :+1:

Reply:
> I was hoping we'd be some caching for free

I think we do get caching for free as long as we can use `getEntityRecords` all the time for `findTemplate`. The current problem is that `getEntityRecord` cannot query for a template by path, so we are stuck doing our own ad-hoc request for it :(

> I truly don't mean to stall this PR. Maybe we should remove the caching for now and add it separately in a subsequent PR?

This sounds perfectly acceptable to me! Given that we want to spend more time thinking about it, it deserves its own PR.

Reply:
How do you feel about just returning the resolved ID from `setPage`? it works well and doesn't really compromise on separation of concerns

Reply:
WFM!